File: 56.merge-intervals.py
# @leet start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        result = []

        left = intervals[0][0]
        right = intervals[0][1]

        for i in range(1, len(intervals)):
            v_left, v_right = intervals[i][0], intervals[i][1]

            start =  max(left, v_left)
            end = min(right, v_right)
            if start <= end:
                left = min(left, v_left)
                right = max(right, v_right)
            else:
                result.append([left, right])
                left = v_left
                right = v_right

        result.append([left, right])

        return result
        
# @leet end

File: 56.merge-intervals.cpp
// @leet start
bool comp(vector<int> a, vector<int> b) {
    if (a[0] == b[0]) {
        return a[1] < b[1];
    }
    return a[0] < b[0];
}

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), comp);
        vector<vector<int>> result;
        vector<int> curr = intervals[0];

        for (int i = 0; i < intervals.size(); i++) {
            vector<int> inter = intervals[i];
            if (curr[0] <= inter[0] && curr[1] >= inter[0]) {
                curr[1] = max(curr[1], inter[1]);
            } else {
                result.push_back(curr);
                curr = inter;
            }
        }

        result.push_back(curr);

        return result;
    }
};
// @leet end
